experiment_api/metric.py

```diff
@@ -0,0 +1,264 @@
+""" Run metrics to determine how well the ResNet model identify the same car
+
+This script will create a plot that will have
+ * x-axis: the range for the number of times we have to go through in the set sorted by the shortest
+           cosine distance to match the target car image
+ * y-axis: cummulative sum of
+       (the number of times we have to go through in the set to match the target car image)
+       divided by
+       (the number of run)
+
+Reliance:
+    The script will rely on ExperimentGenerator to build a list of sets (of images) for the experiment.
+
+Metric:
+    1. CMC: compare the target car with a set of 10 car images
+       Step #1: Using the ExperimentGenerator, create a set of 10 car images with zero drop rate,
+                meaning the target car will exist in the set.
+       Step #2: Match each feature vector to the images in the set.
+       Step #3: Calculate the cosine distance of the target car's feature vector and the feature vector
+                of each image in the set.
+       Step #4: Figure out which image has the shortest cosine distance to the target car.
+       Step #5: Sort the list of images in the set by the shortest cosine distance.
+       Step #6: Determine how many images it has to go through in the set to match the target car.
+                Add this value into a list.
+       Step #7: Plot out the graph.
+
+    2. STR: compare a set of 10 car images with another set of 10 car images
+       Similar to CMC, except that 10 car images are compared to another 10 car images instead of
+       the target car image compared to 10 car images.
+
+Usage:
+    metric.py [-hv]
+    metric.py [-e <SEED> -y <TYPE> -l -c -s] -r <NUM_RUN> <VERI> <FEATURE>
+
+Arguments:
+    VERI                            : Path to the VeRi dataset unzipped
+    FEATURE                         : Path to the feature json file
+                                      Make sure that feature is the same type as TYPE
+
+Options:
+    -h, --help                      : Show this help message.
+    -v, --version                   : Show the version number.
+    -e, --seed=<SEED>               : Seed to be used for random number generator (default: random [1-100])
+    -y, --type=<TYPE>               : Determine which type of images and features to use.
+                                      0: all, 1: query, 2: test, 3: train (default: test)
+    -r, --num_run=<NUM_RUN>         : How many iterations to run the ranking
+    -l, --log                       : Flag to output debug messages (default: info messages)
+    -c, --cmc                       : Run CMC metric
+    -s, --str                       : Run STR metric
+
+"""
+
+import collections
+import docopt
+import logging
+import matplotlib.pyplot as plt
+import os
+import random
+import scipy
+import scipy.spatial
+import sys
+
+from experiment import ExperimentGenerator
+from utils import *
```

Comment:
`import *` is generally frowned upon; prefer `import utils` and then `utils.foo()` or `from utils import foo(), bar(), baz()`.